### PR TITLE
Adds http integration tests 

### PR DESCRIPTION
**Pull request by codefromthecrypt**
Many thanks to Kamon https://github.com/kamon-io/kamon-play/blob/master/kamon-play-2.6.x/src/test/scala/kamon/play/RequestInstrumentationSpec.scala as I used some approaches there to weld brave tests into this. It isn't a great job of it, but starting to work.

this test fails as the current span is not in scope, making the child a part of a different trace. I'm not sure where the scoping of the span is missing

### CODE REVIEW

**Comment by adriancole** [comment]
@takezoe will need a hand to get this one complete

**Comment by adriancole** [comment]
ps I have no idea how to get the executor working like I want... basically I want the test routes to use an executor like below.. Right now, all the routes cannot see the current span

```scala

class CurrentTraceContextExecutionContext(currentTraceContext: CurrentTraceContext, delegate: ExecutionContext) extends ExecutionContext {
  override def execute(runnable: Runnable) = delegate.execute(runnable)

  override def prepare() = new ExecutionContext {
    val invocationContext = currentTraceContext.get()

    def execute(r: Runnable) = delegate.execute(() => {
      val scope = currentTraceContext.newScope(invocationContext)
      try {
        r.run()
      } finally {
        scope.close()
      }
    })

    override def reportFailure(cause: Throwable) = delegate.reportFailure(cause)
  }

  override def reportFailure(cause: Throwable) = delegate.reportFailure(cause)
}
```

**Comment by adriancole** [comment]
I saw this, which might be helpful, just would be nice to inject the Tracing component into it vs looking up the current one each time https://gist.github.com/yanns/f7da61c582ab1da3535d

**Comment by adriancole** [comment]
@takezoe FYI I think this code is ready for review. What we do about implicit trace context propagation can be a separate change

**Comment by takezoe** [comment]
@adriancole Sorry for my slow response. I always thank for your great work! I'll look at this pull request this weekend.

**Comment by adriancole** [comment]
thanks @takezoe I added a commit which incorporate's brave's http.route based naming. This doesn't work because I currently can't read the route.

File: play-zipkin-tracing/play26/src/main/scala/jp/co/bizreach/trace/play26/filter/ZipkinTraceFilter.scala [hunk 1]
```diff
@@ -27,31 +32,64 @@ import scala.util.Failure
...
   def apply(nextFilter: (RequestHeader) => Future[Result])(req: RequestHeader): Future[Result] = {
-    val serverSpan = tracer.serverReceived(
-      spanName = reqHeaderToSpanName(req),
-      span = tracer.newSpan(req.headers)((headers, key) => headers.get(key))
-    )
-    val result = nextFilter(req.withHeaders(new Headers(
-      (req.headers.toMap.mapValues(_.headOption getOrElse "") ++ tracer.toMap(serverSpan)).toSeq
-    )))
+    var template = req.attrs.get(Router.Attrs.HandlerDef).map(_.path)
+    // TODO ^^ is always None! is it being read too soon?
```
Comment (takezoe) [inline play-zipkin-tracing/play26/src/main/scala/jp/co/bizreach/trace/play26/filter/ZipkinTraceFilter.scala]: `HandlerDef` is supplied by Play from the parsed routes file. If we use a custom router, it isn't unavailable.

**Comment by adriancole** [comment]
Oh that is very interesting. Maybe i can change this to reading a literal
for the routes file? Unless there is some way to simulate that otherwise.

Very good find

On 25 Feb 2018 2:26 pm, "Naoki Takezoe" <notifications@github.com> wrote:

*@takezoe* commented on this pull request.
------------------------------

In play-zipkin-tracing/play26/src/main/scala/jp/co/bizreach/
trace/play26/filter/ZipkinTraceFilter.scala
<https://github.com/bizreach/play-zipkin-tracing/pull/32#discussion_r170441758>
:

> @@ -27,31 +32,64 @@ import scala.util.Failure
 class ZipkinTraceFilter @Inject() (tracer:
ZipkinTraceServiceLike)(implicit val mat: Materializer) extends Filter
{

   import tracer.executionContext
-  private val reqHeaderToSpanName: RequestHeader => String =
ZipkinTraceFilter.ParamAwareRequestNamer
+
+  // TODO: any way to get rid of this asInstanceOf?
+  private val handler: HttpServerHandler[RequestHeader, Result] =

It seems to be handled as HttpServerHandler[_, _] because the result type
of HttpServerHandler.create() does not have type parameters. I sent a pull
request to brave to fix it: openzipkin/brave#622
<https://github.com/openzipkin/brave/pull/622>
------------------------------

In play-zipkin-tracing/play26/src/main/scala/jp/co/bizreach/
trace/play26/filter/ZipkinTraceFilter.scala
<https://github.com/bizreach/play-zipkin-tracing/pull/32#discussion_r170443207>
:

>
   def apply(nextFilter: (RequestHeader) => Future[Result])(req:
RequestHeader): Future[Result] = {
-    val serverSpan = tracer.serverReceived(
-      spanName = reqHeaderToSpanName(req),
-      span = tracer.newSpan(req.headers)((headers, key) => headers.get(key))
-    )
-    val result = nextFilter(req.withHeaders(new Headers(
-      (req.headers.toMap.mapValues(_.headOption getOrElse "") ++
tracer.toMap(serverSpan)).toSeq
-    )))
+    var template = req.attrs.get(Router.Attrs.HandlerDef).map(_.path)
+    // TODO ^^ is always None! is it being read too soon?

HandlerDef is supplied by Play from the parsed routes file. If we use a
custom router, it isn't unavailable.

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub
<https://github.com/bizreach/play-zipkin-tracing/pull/32#pullrequestreview-99132109>,
or mute the thread

**Comment by adriancole** [comment]
@simao sorry I never finished this but it would have solved the two pull requests you have

**Comment by simao** [comment]
@adriancole Ok, so then what is the status of this and how could I help move this forward?

File: play-zipkin-tracing/play26/src/main/scala/jp/co/bizreach/trace/play26/filter/ZipkinTraceFilter.scala (hunk 1 above)
Comment (adriancole) [reply play-zipkin-tracing/play26/src/main/scala/jp/co/bizreach/trace/play26/filter/ZipkinTraceFilter.scala]: @simao last place waiting for answer was here. Do you know how we can handle custom router? We should ideally have tests to cover both routes file and custom router

**Pull request closed without merging**